File: app/assets/javascripts/cqm/StatementResult.js
const mongoose = require('mongoose/browser');

const [String, Mixed] = [
  mongoose.Schema.Types.String,
  mongoose.Schema.Types.Mixed,
];

const StatementResultSchema = mongoose.Schema({
  // Library the statement this result is for is in
  library_name: String,
  // Statement this result is for is in
  statement_name: String,
  // Result, processed for display, of the statement this result is for
  pretty: String,
  // Final, processed result of raw calculation
  final: String,
  // Raw result of clause calculation
  raw: Mixed,
  /*
   * 'NA' - Not applicable. This statement is not relevant to any population calculation in this population_set. Common
   *   for unused library statements or statements only used for other population sets.
   *
   * 'FALSE' - This statement is not relevant to any of this patient's population inclusion calculations.
   *
   * 'TRUE' - This statement is relevant for one or more of the population inclusion calculations.
    */
  relevance: {
    type: String,
    enum: ['NA', 'TRUE', 'FALSE'],
    default: 'NA',
  },
});


module.exports.StatementResultSchema = StatementResultSchema;
class StatementResult extends mongoose.Document {
  constructor(object) {
    super(object, StatementResultSchema);
  }
}
module.exports.StatementResult = StatementResult;

File: app/models/cqm/statement_result.rb
module CQM
  # StatementResult is used to store statement result information in IndividualResult
  class StatementResult
    include Mongoid::Document
    include Mongoid::Timestamps

    # These are the Mongoid equivalent of the Mongoose "enum" attribute for the respective fields.
    # They throw an error if you try to assign a value that's not in the array.
    validates_inclusion_of :relevance, in: %w[NA TRUE FALSE]

    # Library the clause this result is for is in
    field :library_name, type: String
    # Statment the clause this result is for is in
    field :statement_name, type: String
    # LocalId of the clause this result is for
    field :pretty, type: String
    # Final, processed result of raw calculation
    field :final, type: String
    # Raw result of clause calculation
    field :raw

    field :relevance, type: String

    # Relations to other model classes
    embedded_in :individual_result
  end
end
